src/lib.rs

```diff
@@ -301,3 +302,262 @@ impl GapfillDeltaTransition {
         return self.deltas.clone();
     }
 }
+
+//a vector selector aggregate has the same semantics as parse.VectorSelector processing
+//in Prometheus. Namely, for all timestamps ts in the series:
+//     ts := ev.startTimestamp; ts <= ev.endTimestamp; ts += ev.interval
+// we return the last sample.value such that the time sample.time <= ts and sample.time >= ts-lookback.
+// if such a sample doesn't exist return NULL (None).
+// thus, the vector selector returns a regular series of values corresponding to all the points in the
+// ts series above.
+// Note that for performance, this aggregate is parallel-izable, combinable, and does not expect ordered inputs.
+#[pg_extern(immutable, parallel_safe)]
+pub fn vector_selector_transition(
+    state: Option<Internal<VectorSelector>>,
+    start_time: TimestampTz,
+    end_time: TimestampTz,
+    bucket_width: Milliseconds,
+    lookback: Milliseconds,
+    time: TimestampTz,
+    val: f64,
+    fc: pg_sys::FunctionCallInfo,
+) -> Option<Internal<VectorSelector>> {
+    unsafe {
+        in_aggregate_context(fc, || {
+            let mut state = state.unwrap_or_else(|| {
+                let state: Internal<_> =
+                    VectorSelector::new(start_time, end_time, bucket_width, lookback).into();
+                state
+            });
+
+            state.insert(time, val);
+
+            Some(state)
+        })
+    }
+}
+
+#[pg_extern(immutable, parallel_safe)]
+pub fn vector_selector_final(state: Option<Internal<VectorSelector>>) -> Option<Vec<Option<f64>>> {
+    state.map(|s| s.to_pg_array())
+}
+
+#[allow(non_camel_case_types)]
+type bytea = pg_sys::Datum;
+
+#[pg_extern(immutable, parallel_safe)]
+pub fn vector_selector_serialize(state: Internal<VectorSelector>) -> bytea {
+    crate::do_serialize!(state)
+}
+
+#[pg_extern(immutable, parallel_safe)]
+pub fn vector_selector_deserialize(
+    bytes: bytea,
+    _internal: Option<Internal<()>>,
+) -> Internal<VectorSelector> {
+    crate::do_deserialize!(bytes, VectorSelector)
+}
+
+#[pg_extern(immutable, parallel_safe)]
+pub fn vector_selector_combine(
+    state1: Option<Internal<VectorSelector>>,
+    state2: Option<Internal<VectorSelector>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Internal<VectorSelector>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            match (state1, state2) {
+                (None, None) => None,
+                (None, Some(state2)) => {
+                    let s = state2.clone();
+                    Some(s.into())
+                }
+                (Some(state1), None) => {
+                    let s = state1.clone();
+                    Some(s.into())
+                } //should I make these return themselves?
+                (Some(state1), Some(state2)) => {
+                    let mut s1 = state1.clone(); // is there a way to avoid if it doesn't need it
+                    s1.combine(&state2);
```

Comment:
if you make `combine()` take `&self` and be in charge of allocating you can remove the clone outside it.